iroh-sync/src/store/fs.rs

```diff
@@ -341,73 +390,70 @@ impl Store {
             return Err(anyhow!("replica is not closed"));
         }
-        let write_tx = self.db.begin_write()?;
-        {
-            let mut record_table = write_tx.open_table(RECORDS_TABLE)?;
+        self.modify(|tables| {
             let bounds = RecordsBounds::namespace(*namespace);
-            record_table.retain_in(bounds.as_ref(), |_k, _v| false)?;
-        }
-        {
-            let mut table = write_tx.open_table(RECORDS_BY_KEY_TABLE)?;
+            tables.records.retain_in(bounds.as_ref(), |_k, _v| false)?;
             let bounds = ByKeyBounds::namespace(*namespace);
-            let _ = table.retain_in(bounds.as_ref(), |_k, _v| false);
-        }
-        {
-            let mut namespace_table = write_tx.open_table(NAMESPACES_TABLE)?;
-            namespace_table.remove(namespace.as_bytes())?;
-        }
-        {
-            let mut peers_table = write_tx.open_multimap_table(NAMESPACE_PEERS_TABLE)?;
-            peers_table.remove_all(namespace.as_bytes())?;
-            let mut dl_policies_table = write_tx.open_table(DOWNLOAD_POLICY_TABLE)?;
-            dl_policies_table.remove(namespace.as_bytes())?;
-        }
-        write_tx.commit()?;
-        Ok(())
+            let _ = tables
+                .records_by_key
+                .retain_in(bounds.as_ref(), |_k, _v| false);
+            tables.namespaces.remove(namespace.as_bytes())?;
+            tables.namespace_peers.remove_all(namespace.as_bytes())?;
+            tables.download_policy.remove(namespace.as_bytes())?;
+            Ok(())
+        })
     }
 
     /// Get an iterator over entries of a replica.
     pub fn get_many(
-        &self,
+        &mut self,
         namespace: NamespaceId,
         query: impl Into<Query>,
     ) -> Result<QueryIterator> {
-        let read_tx = self.db.begin_read()?;
-        QueryIterator::new(&read_tx, namespace, query.into())
+        QueryIterator::new(self.tables()?, namespace, query.into())
     }
 
     /// Get an entry by key and author.
     pub fn get_exact(
-        &self,
+        &mut self,
         namespace: NamespaceId,
         author: AuthorId,
         key: impl AsRef<[u8]>,
         include_empty: bool,
     ) -> Result<Option<SignedEntry>> {
-        let read_tx = self.db.begin_read()?;
-        let record_table = read_tx.open_table(RECORDS_TABLE)?;
-        get_exact(&record_table, namespace, author, key, include_empty)
+        get_exact(
+            &self.tables()?.records,
+            namespace,
+            author,
+            key,
+            include_empty,
+        )
     }
 
     /// Get all content hashes of all replicas in the store.
-    pub fn content_hashes(&self) -> Result<ContentHashesIterator> {
-        let read_tx = self.db.begin_read()?;
-        ContentHashesIterator::new(&read_tx)
+    pub fn content_hashes(&mut self) -> Result<ContentHashesIterator> {
+        // make sure the current transaction is committed
+        self.flush()?;
```

Comment:
QueryIterator has a lifetime and only needs a reference to a Table<a'>. ContentHashesIterator needs a *copy* of a table.

Reply:
Ah, this is because we are sending the `ContentHashesIterator` over the channel boundary directly, whereas the QueryIterator is iterated within the actor, and each item pushed into a flume channel. Correct?
Maybe you can add a comment somewhere that mentions these two different approaches.
Did you think about what the perfomance implications are for each approach?

Reply:
> Ah, this is because we are sending the ContentHashesIterator over the channel boundary directly, whereas the QueryIterator is iterated within the actor, and each item pushed into a flume channel. Correct?

Yes.

If the QueryIterator was an owned iterator it would have to be created on a read transaction, since only ReadOnlyTables are without lifetimes. So the iterator would be the same speed, but you might have to constantly switch between read and write transactions.

For GC I think making a snapshot is OK, but if you would make one for each query you would defeat the purpose of the batching.